pulp_smash/tests/puppet/api_v2/test_sync_publish.py

```diff
@@ -0,0 +1,655 @@
+            'upload',
+            'import',
+            'upload_free',
+            'copy',
+            'publish',
+            'distribute',
+            'search units',
+            'query units',
+        }
+        cls.responses = {key: [] for key in steps}
+        cls.bodies = {}
+        cls.original_puppets = []  # binary blobs
+
+        # Download puppet modules.
+        response = requests.get(_PUPPET_QUERY_URL)
+        response.raise_for_status()
+        cls.original_puppets.append(response.content)
+
+        # Create two repos, start upload request, upload module to upload dir
+        cls.attrs_iter = tuple((
+            create_repository(cls.cfg, _gen_puppet_repo_body())
+            for _ in range(2)
+        ))
+        cls.bodies['upload_malloc'] = _start_content_upload(
+            cls.cfg,
+            cls.responses['upload_malloc']
+        )
+        cls.bodies['upload'] = _upload_file(
+            cls.cfg,
+            cls.bodies['upload_malloc']['_href'],
+            cls.original_puppets[0],
+            cls.responses['upload'],
+        )
+
+        # Import uploaded puppet modules into first repo. Poll.
+        # Delete upload request.
+        cls.bodies['import'] = _import_puppet_to_repo(
+            cls.cfg,
+            cls.bodies['upload_malloc']['upload_id'],
+            cls.attrs_iter[0]['_href'],
+            cls.responses['import'],
+        )
+        for call_report in cls.bodies['import']:
+            poll_spawned_tasks(cls.cfg, call_report)
+        cls.bodies['upload_free'] = delete(
+            cls.cfg,
+            cls.bodies['upload_malloc']['_href'],
+            cls.responses['upload_free'],
+        )
+
+        # Copy content from first repo to second. Poll until done.
+        cls.bodies['copy'] = _copy_repo(
+            cls.cfg,
+            cls.attrs_iter[0]['id'],
+            cls.attrs_iter[1]['_href'],
+            cls.responses['copy'],
+        )
+        for call_report in cls.bodies['copy']:
+            poll_spawned_tasks(cls.cfg, call_report)
+
+        # Add a distributor to the repositories, then publish repositories
+        # to the distributors. Poll the publish request.
+        cls.bodies['distribute'] = tuple((
+            _add_puppet_distributor(
+                cls.cfg,
+                cls.attrs_iter[i]['_href'],
+                cls.responses['distribute']
+            ) for i in range(2)
+        ))
+        cls.bodies['publish'] = tuple((
+            _publish_repo(
+                cls.cfg,
+                cls.attrs_iter[i]['_href'],
+                cls.bodies['distribute'][i]['id'],
+                cls.responses['publish'],
+            ) for i in range(2)
+        ))
+        for i in range(2):
+            for call_report in cls.bodies['publish'][i]:
+                poll_spawned_tasks(cls.cfg, call_report)
+
+        # Query modules on server. Data integrity is tested in tests below.
+        query_href = '/v3/releases?module={}/{}'.format(
```

Comment:
We need to test that you can use the older versions of the puppet forge API (puppet clients prior to version 3.3).

On my dev environment I am able to run:

http --verify=no GET http://.:repo1@hostname/api/v1/releases.json?module=domcleal/katellovirt

In response I get json that tells me where that module can be downloaded. The basic auth in the above request is explained here [0].

We should also attempt to query the forge API used for clients between versions 3.3 and 3.6. The way the request is formulated is described here [1]
These two additional scenarios will work on Pulp 2.6+, however the test that is already implemented will only work for Pulp version 2.7+.

[0] http://pulp-puppet.readthedocs.org/en/latest/tech-reference/forge_api.html#basic-auth
[1] http://pulp-puppet.readthedocs.org/en/latest/user-guide/recipes.html#installing-with-puppet-client-3-3